iceoryx_examples/automotive_soa/src/runtime.cpp

```diff
@@ -0,0 +1,173 @@
+        {
+            break;
+        }
+    }
+    if (iter != m_callbacks.end())
+    {
+        m_callbacks.erase(iter);
+    }
+
+    if (m_callbacks.empty())
+    {
+        m_listener.detachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED);
+    }
+}
+
+bool Runtime::verifyThatServiceIsComplete(kom::ServiceHandleContainer<kom::ProxyHandleType>& container) noexcept
+{
+    // The service level of AUTOSAR Adaptive is not available in iceoryx, instead every publisher and server is
+    // considered as a service. A ara::com binding implementer would typically query the AUTOSAR meta model here, to
+    // find out if all event, fields and methods of a service are available. For the example we assume that the
+    // 'MinimalSkeleton' service is complete when the container contains the four iceoryx services:
+    //
+    // 1.    EventPublisher: MinimalSkeleton, Instance, Event  (iox::popo::Publisher)
+    // 2. a) FieldPublisher: MinimalSkeleton, Instance, Field  (iox::popo::Publisher)
+    // 2. b) FieldPublisher: MinimalSkeleton, Instance, Field  (iox::popo::Server)
+    // 3.    MethodServer:   MinimalSkeleton, Instance, Method (iox::popo::Server)
+
+    if (container.size() == 4U)
+    {
+        return true;
+    }
+    return false;
+}
+
+void Runtime::invokeCallback(iox::runtime::ServiceDiscovery*, Runtime* self) noexcept
+{
+    // Has the availability of one of the registered services changed?
+    //! [perform FindService]
+    for (auto& callback : self->m_callbacks)
+    {
+        auto container =
+            self->FindService(std::get<1>(callback).m_serviceIdentifier, std::get<1>(callback).m_instanceIdentifier);
+
+        auto numberOfAvailableServicesOnCurrentSearch = container.size();
+        auto& numberOfAvailableServicesOnLastSearch = std::get<2>(callback);
+        //! [perform FindService]
+
+        auto executeCallback = [&]() {
+            (std::get<0>(callback))(container,
+                                    kom::FindServiceHandle({std::get<1>(callback).m_serviceIdentifier,
+                                                            std::get<1>(callback).m_instanceIdentifier}));
+            numberOfAvailableServicesOnLastSearch.emplace(numberOfAvailableServicesOnCurrentSearch);
+        };
+
+        // If the service is available for the first time
+        //! [first execution conditions]
+        if (!numberOfAvailableServicesOnLastSearch.has_value() && numberOfAvailableServicesOnCurrentSearch != 0)
+        {
+            executeCallback();
+        }
+        //! [first execution conditions]
+
+        // If the service was available before and current number of services has changed
+        //! [second execution conditions]
+        if (numberOfAvailableServicesOnLastSearch.has_value()
+            && numberOfAvailableServicesOnLastSearch.value() != numberOfAvailableServicesOnCurrentSearch)
+        {
+            executeCallback();
```

Comment:
I think this cannot happen because checking `numberOfAvailableServicesOnLastSearch.has_value()` makes the conditions exclusive.

Reply:
I can second what @MatthiasKillat wrote, I didn't saw that behaviour because with `numberOfAvailableServicesOnLastSearch.has_value()` it is ensured we react accordingly with either the first or second execution condition.